sonar-text-plugin/src/main/java/org/sonar/plugins/common/DurationStatistics.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * SonarQube Text Plugin
+ * Copyright (C) 2021-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.common;
+
+import java.text.DecimalFormat;
+import java.text.DecimalFormatSymbols;
+import java.text.Format;
+import java.text.NumberFormat;
+import java.util.Comparator;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonar.api.config.Configuration;
+
+public class DurationStatistics {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DurationStatistics.class);
+
+  private static final String PROPERTY_KEY = "sonar.text.duration.statistics";
+
+  public static final String SUFFIX_TOTAL = "::total";
+  public static final String SUFFIX_PRE = "::preFilter";
+  public static final String SUFFIX_MATCHER = "::matcher";
+  public static final String SUFFIX_POST = "::postFilter";
+
+  private final Map<String, Measurement> stats = new ConcurrentHashMap<>();
+
+  private final AtomicBoolean isRecordingEnabled = new AtomicBoolean(false);
+
+  public DurationStatistics(Configuration config) {
+    config.getBoolean(PROPERTY_KEY).ifPresent(isRecordingEnabled::set);
+  }
+
+  public <T> T timed(String id, Supplier<T> supplier) {
+    if (isRecordingEnabled.get()) {
+      long startTime = System.nanoTime();
+      var result = supplier.get();
+      addRecord(id, System.nanoTime() - startTime);
+      return result;
+    } else {
+      return supplier.get();
+    }
+  }
+
+  void addRecord(String id, long elapsedTime) {
+    stats.computeIfAbsent(id, key -> new Measurement()).add(elapsedTime);
+  }
+
+  public void log() {
+    if (isRecordingEnabled.get()) {
+      var symbols = new DecimalFormatSymbols(Locale.ROOT);
+      symbols.setGroupingSeparator('\'');
+      NumberFormat format = new DecimalFormat("#,###", symbols);
```

Comment:
Maybe this should be moved to `formatEntry` or to class level as we always use the same formatting?